src/feature/hs/hs_circuit.c

```diff
@@ -958,6 +959,42 @@ hs_circ_handle_intro_established(const hs_service_t *service,
   return ret;
 }
 
+/**
+ *  Go into <b>data</b> and add the right subcredential to be able to handle
+ *  this incoming cell.
+ *
+ *  <b>desc_subcred</b> is the subcredential of the descriptor that corresponds
+ *  to the intro point that received this intro request. This subcredential
+ *  should be used if we are not an onionbalance instance.
+ *
+ *  Return 0 if everything went well, or -1 in case of internal error.
+ */
+static int
+get_subcredential_for_handling_intro2_cell(const hs_service_t *service,
+                                        hs_cell_introduce2_data_t *data,
+                                        const hs_subcredential_t *desc_subcred)
+{
+  /* Handle the simple case first: We are not an onionbalance instance and we
+   * should just use the regular descriptor subcredential */
+  if (!hs_ob_service_is_instance(service)) {
+    data->n_subcredentials = 1;
+    data->subcredentials = desc_subcred;
+    return 0;
+  }
+
+  /* This should not happen since we should have made onionbalance
+   * subcredentials when we created our descriptors. */
+  if (BUG(!service->ob_subcreds)) {
+    return -1;
+  }
+
+  /* We are an onionbalance instance: */
```

Comment:
ob_subreds also includes the descriptor's subcredentials as seen in `compute_subcredentials`.

I agree that the function could be simpler that way, but I'm not sure how to do this while preserving the optimal performance of the non-OB mode. In particular, when the INTRO2 is received in the non-OB case, the service can know the exact subcredential that should be used because each intro point corresponds to a specific descriptor and hence to a specific subcredential. So in the non-OB case, we don't even need to try both descriptor subcredentials, we can only try one.

Reply:
ah, ok.  This seems like something we should look at in a ticket down the road then; maybe we could have a list of subcredentials for each intro point?  But it's not so bad to do things this way if there's a reason.